### PR TITLE
kubeadm: allow imageName to be specified in kubeadm configuration

### PR DESCRIPTION
**Pull request by thtanaka**
The ability to specify a specific image name to be used for kubeadm config
is useful in case other components that kubeadm depended on decided to
rename its image.

Signed-off-by: Thomas Tanaka <thomas.tanaka@oracle.com>

#### What type of PR is this?
/kind feature
/kind regression

#### What this PR does / why we need it:
The ability to specify an image name will be useful as other components
that kubeadm do not have a direct control of might decided to change its
image name, e.g in k8s 1.21.1 coredns image is named as "coredns/coredns".

#### Does this PR introduce a user-facing change?
```
release-note

```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:
```
docs

```

### CODE REVIEW

File: cmd/kubeadm/app/apis/kubeadm/v1beta2/doc.go [hunk 1]
```diff
@@ -188,6 +188,7 @@ limitations under the License.
 // 	  local:
 // 	    imageRepository: "k8s.gcr.io"
 // 	    imageTag: "3.2.24"
+//	    imageName: "etcd"
```
Comment (neolit123) [inline cmd/kubeadm/app/apis/kubeadm/v1beta2/doc.go:191]: v1beta2 is locked and no more changes should be done to it.

File: cmd/kubeadm/app/apis/kubeadm/v1beta2/types.go [hunk 2]
```diff
@@ -163,7 +163,9 @@ type ImageMeta struct {
 	// In case this value is set, kubeadm does not change automatically the version of the above components during upgrades.
 	ImageTag string `json:"imageTag,omitempty"`
 
-	//TODO: evaluate if we need also a ImageName based on user feedbacks
+	// ImageName allows to specify an exact name of the image.
+	// if not set, the ImageName defined in a constants.go will be used instead.
+	ImageName string `json:"imageName,omitempty"`
```
Comment (neolit123) [inline cmd/kubeadm/app/apis/kubeadm/v1beta2/types.go:168]: v1beta2 is locked and no more changes should be done to it.

**Comment by neolit123** [comment]
hello, there is a discussion on https://github.com/kubernetes/kubeadm/issues/2411 about this.
in fact, we wanted to not make this change because it makes upgrades difficult.

in custom registries you can name your components `kube-apiserver` and use custom tags such as `kube-apiserver:v1.21.0-foo`.
do you have an use case for this change or were you following the TODO?
i think we might want to just remove the TODO from v1beta2 | 3.

/cc @fabriziopandini 
/hold

**Comment by thtanaka** [comment]
@neolit123 thanks a lot for your review, yes I was actually hitting this problem when trying a stable kubernetes 1.21.1 which has a `coredns` image defined as `coredns/coredns`. I hit this issue since our internal build of `coredns` is still tagged with `coredns` vs `coredns/coredns`. While debugging this problem, I just come across the `TODO` and decided that this could be a useful feature.

**Comment by fabriziopandini** [comment]
When it comes to API changes, my suggestion is to always start a discussion on a issue and reach agreement before moving on with the implementation.

With regards to root cause for the problem described above, we agreed that using `coredns/coredns` as image name was a regression, and thus we already merged https://github.com/kubernetes/kubernetes/pull/102502 and back ported the change for the upcoming v1.21.2 release.

TL;DR; when using k8s.gcr.io, kubeadm will take care of adding the subpath, thus aligning on the change of the upstream CoreDNS image; instead, when using a custom image repository kubeadm will append the "coredns"  image name instead of  "coredns/coredns", thus restoring the behaviour existing before the v1.21 release. See release notes on the PR for more context

**Comment by thtanaka** [comment]
I will take note of that. Thank you for your review.
